File: src/Crawler/Analysis/Result/AnalyzerStats.php
<?php

/*
 * This file is part of the SiteOne Crawler.
 *
 * (c) Ján Regeš <jan.reges@siteone.cz>
 */

declare(strict_types=1);

namespace Crawler\Analysis\Result;

class AnalyzerStats
{

    /**
     * @var array <string, array<string, array>>
     */
    private array $severityCountsPerAnalysis = [];

    /**
     * Add OK result to stats. You can optionally specify subject (e.g. URL or SVG body) to count only unique results.
     *
     * @param string $analysisName
     * @param string|null $subject
     * @return void
     */
    public function addOk(string $analysisName, ?string $subject = null): void
    {
        $this->addResult($analysisName, 'ok', $subject);
    }

    /**
     * Add WARNING result to stats. You can optionally specify subject (e.g. URL or SVG body) to count only unique results.
     *
     * @param string $analysisName
     * @param string|null $subject
     * @return void
     */
    public function addWarning(string $analysisName, ?string $subject = null): void
    {
        $this->addResult($analysisName, 'warning', $subject);
    }

    /**
     * Add CRITICAL result to stats. You can optionally specify subject (e.g. URL or SVG body) to count only unique results.
     *
     * @param string $analysisName
     * @param string|null $subject
     * @return void
     */
    public function addCritical(string $analysisName, ?string $subject = null): void
    {
        $this->addResult($analysisName, 'critical', $subject);
    }

    /**
     * Add NOTICE result to stats. You can optionally specify subject (e.g. URL or SVG body) to count only unique results.
     *
     * @param string $analysisName
     * @param string|null $subject
     * @return void
     */
    public function addNotice(string $analysisName, ?string $subject = null): void
    {
        $this->addResult($analysisName, 'notice', $subject);
    }

    /**
     * @return array <int<0, max>, array<string, int<0, max>|string>>
     */
    public function toTableArray(): array
    {
        $result = [];
        foreach ($this->severityCountsPerAnalysis as $analysisName => $severityCounts) {
            $result[] = [
                'analysisName' => $analysisName,
                'ok' => count($severityCounts['ok']),
                'notice' => count($severityCounts['notice']),
                'warning' => count($severityCounts['warning']),
                'critical' => count($severityCounts['critical']),
            ];
        }
        return $result;
    }

    private function addResult(string $analysisName, string $severity, ?string $subject): void
    {
        if (!isset($this->severityCountsPerAnalysis[$analysisName])) {
            $this->severityCountsPerAnalysis[$analysisName] = [
                'ok' => [],
                'warning' => [],
                'critical' => [],
                'notice' => [],
            ];
        }

        // subject hash - 10 chars should be enough to identify unique subjects (1,099,511,627,776 combinations)
        $subjectHash = $subject !== null ? substr(md5(trim($subject)), 0, 10) : null;

        if ($subjectHash) {
            $this->severityCountsPerAnalysis[$analysisName][$severity][$subjectHash] = true;
        } else {
            $this->severityCountsPerAnalysis[$analysisName][$severity][] = true;
        }
    }

}

File: src/analysis/result/analyzer_stats.rs
// SiteOne Crawler - AnalyzerStats
// (c) Jan Reges <jan.reges@siteone.cz>

use std::collections::HashMap;

#[derive(Debug, Clone, Default)]
pub struct AnalyzerStats {
    /// analysis_name -> severity -> set of subject hashes (or just counted entries)
    severity_counts_per_analysis: HashMap<String, SeverityCounts>,
}

#[derive(Debug, Clone, Default)]
struct SeverityCounts {
    ok: HashMap<String, bool>,
    notice: HashMap<String, bool>,
    warning: HashMap<String, bool>,
    critical: HashMap<String, bool>,
}

impl AnalyzerStats {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_ok(&mut self, analysis_name: &str, subject: Option<&str>) {
        self.add_result(analysis_name, "ok", subject);
    }

    pub fn add_warning(&mut self, analysis_name: &str, subject: Option<&str>) {
        self.add_result(analysis_name, "warning", subject);
    }

    pub fn add_critical(&mut self, analysis_name: &str, subject: Option<&str>) {
        self.add_result(analysis_name, "critical", subject);
    }

    pub fn add_notice(&mut self, analysis_name: &str, subject: Option<&str>) {
        self.add_result(analysis_name, "notice", subject);
    }

    pub fn to_table_data(&self) -> Vec<HashMap<String, String>> {
        let mut result = Vec::new();
        for (analysis_name, counts) in &self.severity_counts_per_analysis {
            let mut row = HashMap::new();
            row.insert("analysisName".to_string(), analysis_name.clone());
            row.insert("ok".to_string(), counts.ok.len().to_string());
            row.insert("notice".to_string(), counts.notice.len().to_string());
            row.insert("warning".to_string(), counts.warning.len().to_string());
            row.insert("critical".to_string(), counts.critical.len().to_string());
            result.push(row);
        }
        result
    }

    fn add_result(&mut self, analysis_name: &str, severity: &str, subject: Option<&str>) {
        let counts = self
            .severity_counts_per_analysis
            .entry(analysis_name.to_string())
            .or_default();

        let subject_hash = subject.map(|s| {
            use md5::{Digest, Md5};
            let mut hasher = Md5::new();
            hasher.update(s.trim().as_bytes());
            let result = hasher.finalize();
            format!("{:x}", result)[..10].to_string()
        });

        let map = match severity {
            "ok" => &mut counts.ok,
            "notice" => &mut counts.notice,
            "warning" => &mut counts.warning,
            "critical" => &mut counts.critical,
            _ => return,
        };

        if let Some(hash) = subject_hash {
            map.insert(hash, true);
        } else {
            // Use a unique key based on current count
            let key = format!("_auto_{}", map.len());
            map.insert(key, true);
        }
    }
}
